modules/core/src/main/scala/vinyldns/core/task/TaskScheduler.scala

```
@@ -0,0 +1,80 @@
+ * limitations under the License.
+ */
+
+package vinyldns.core.task
+import cats.effect._
+import cats.implicits._
+import fs2._
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.duration.FiniteDuration
+
+// Interface for all Tasks that need to be run
+trait Task {
+  // The name of the task, should be unique / constant and should not change
+  def name: String
+
+  // The amount of time this task is running before it can be reclaimed / considered failed
+  def timeout: FiniteDuration
+
+  // How often to attempt to run the task
+  def runEvery: FiniteDuration
+
+  // Runs the task
+  def run(): IO[Unit]
+}
+
+object TaskScheduler {
+  private val logger = LoggerFactory.getLogger("TaskScheduler")
+
+  // Starts a task on a schedule, returns a Fiber that can be used to cancel the task
+  def schedule(task: Task, taskRepository: TaskRepository)(
+      implicit t: Timer[IO],
+      cs: ContextShift[IO]): Stream[IO, Unit] = {
+
+    def claimTask(): IO[Option[Task]] = IO.suspend {
+      taskRepository.claimTask(task.name, task.timeout).map {
+        case true =>
+          logger.info(s"""Successfully found and claimed task; taskName="${task.name}" """)
+          Some(task)
+        case false =>
+          logger.info(s"""No task claimed; taskName="${task.name}" """)
+          None
+      }
+    }
+
+    def releaseTask(maybeTask: Option[Task]): IO[Unit] = IO.suspend {
+      maybeTask
+        .map(
+          t =>
+            taskRepository
+              .releaseTask(t.name)
+              .as(logger.info(s"""Released task; taskName="${task.name}" """)))
+        .getOrElse(IO.unit)
+    }
+
+    def runTask(maybeTask: Option[Task]): IO[Unit] = maybeTask.map(_.run()).getOrElse(IO.unit)
+
+    // Acquires a task, runs it, and makes sure it is cleaned up, swallows the error via a log
+    def runOnceSafely(task: Task): IO[Unit] =
+      claimTask().bracket(runTask)(releaseTask).handleError { error =>
```

`.bracket` makes sure that the finalizer (releaseTask) gets called no matter what.

We want to swallow the error as well